src/mame/drivers/dc_atomiswave.cpp

```diff
@@ -0,0 +1,1536 @@
+                 PIC Usage:
+                           Game                                   Sega Part#
+                           ---------------------------------------------------
+                           Extreme Hunting 2 Tournament Edition   317-0445-COM
+
+
+AW-NET Network Board
+--------------------
+
+Sammy
+AM3AJG-01
+LAN PCB
+2003
+SAMLAN Rev: D
+|-----------------------------------|
+| RJ45                              |
+|PULSE                              |
+|        25Mz                       |
+|                                   |
+|      RTL8139CL+*           CN*    |
+|                  315-6310*        |
+|                                   |
+|       L46R                        |
+|-----------------------------------|
+Notes:
+      *          - Denotes those parts are on the other side of the PCB
+      RJ45       - RJ45 network connector
+      PULSE      - Pulse H0011 10/100 LAN Magnetics Module (SOIC16)
+      L46R       - 93C46 compatible 128x8 EEPROM (SOIC8)
+      RTL8139CL+ - Realtek RTL8139CL+ 3.3 volt Single Chip Fast Ethernet Controller with Power Management (QFP120)
+      315-6310   - Sega 315-6310 Custom IC (QFP100)
+      CN         - This connector plugs into the main board
+
+
+Gun Sub Board
+-------------
+
+AM3AGT-02 GUN SUB PCB
+|------------------|
+|CN5  CN4  CN3  CN2|
+|       74HC74     |
+|                  |
+| 74HC74    7CHC74 |
+|      74HC74      |
+|                  |
+|       CN1        |
+|------------------|
+Notes:
+      CN1 - 8 pin connector joining to I/O Expansion Module (which is plugged into main board)
+      CN2 - Gun connection for player 2 trigger and optical
+      CN3 - Gun connection for player 2 pump switch
+      CN4 - Gun connection for player 1 trigger and optical
+      CN5 - Gun connection for player 1 pump switch
+
+
+Other games not dumped
+----------------------
+Miracle Stadium
+
+Cancelled, prototypes known to exists
+----------------------
+Chase 1929
+Force Five
+Kenju
+Premier Eleven
+
+**************************************************************************************************/
+
+#include "emu.h"
+#include "includes/dc_atomiswave.h"
+
+//#include "machine/gunsense.h"
+//#include "emupal.h"
+#include "screen.h"
+#include "speaker.h"
+
+uint64_t atomiswave_state::aw_flash_r(offs_t offset)
+{
+	return (uint64_t)m_awflash->read(offset*8) | (uint64_t)m_awflash->read((offset*8)+1)<<8 | (uint64_t)m_awflash->read((offset*8)+2)<<16 | (uint64_t)m_awflash->read((offset*8)+3)<<24 |
+			(uint64_t)m_awflash->read((offset*8)+4)<<32 | (uint64_t)m_awflash->read((offset*8)+5)<<40 | (uint64_t)m_awflash->read((offset*8)+6)<<48 | (uint64_t)m_awflash->read((offset*8)+7)<<56;
+}
+
+void atomiswave_state::aw_flash_w(offs_t offset, uint64_t data, uint64_t mem_mask)
+{
+	int i;
+	uint32_t addr = offset * 8;
+
+	for (i = 0; i < 8; i++)
+	{
+		if (mem_mask & ((uint64_t)0xff)<< (i*8))
+		{
+			addr += i;
+			break;
+		}
+	}
+
+	data >>= (i*8);
+
+	m_awflash->write(addr, data);
+}
```

Comment:
Would this whole thing be better to do as an 8-bit handler with a unit mask?

Reply:
Haven't touched this part either but yeah good call: `macronix_29l001mc_device` inherits from `intelfsh8_device` which shouldn't need this trampoline at all.